pkg/bootstrap/bootstrap.go

```diff
@@ -187,54 +247,166 @@ func extractFromDir(dir, prefix string, img v1.Image, imgName string) error {
 			}
 		} else if err != nil {
-			errs = append(errs, errors2.Wrapf(err, "failed to move file %s", src))
+			// Other error while renaming src to dst.
+			errs = append(errs, errors2.Wrapf(err, "failed to rename %q to %q", src, dst))
 		}
 	}
-	if len(errs) > 0 {
-		return merr.NewErrors(errs...)
-	}
-	return nil
+	return merr.NewErrors(errs...)
 }
 
-func preloadBootstrapImage(dataDir, runtimeImage string) (v1.Image, error) {
-	imagesDir := filepath.Join(dataDir, "agent", "images")
+// preloadBootstrapImage attempts return an image named imageName from a tarball
+// within imagesDir.
+func preloadBootstrapImage(dataDir string, imageName string) (v1.Image, error) {
+	imagesDir := imagesDir(dataDir)
 	if _, err := os.Stat(imagesDir); err != nil {
 		if os.IsNotExist(err) {
+			logrus.Debugf("No local image available for %q: dir %q does not exist", imageName, imagesDir)
 			return nil, nil
 		}
 		return nil, err
 	}
+
+	// Walk the images dir to get a list of tar files
 	files := map[string]os.FileInfo{}
 	if err := filepath.Walk(imagesDir, func(path string, info os.FileInfo, err error) error {
 		if err != nil {
 			return err
 		}
-		files[path] = info
+		if !info.IsDir() && strings.HasSuffix(path, ".tar") {
+			files[path] = info
+		}
 		return nil
 	}); err != nil {
 		return nil, err
 	}
-	archTag, err := name.NewTag(runtimeImage, name.WeakValidation)
+
+	imageTag, err := name.NewTag(imageName, name.WeakValidation)
 	if err != nil {
 		return nil, err
 	}
+
+	// Try to find the requested tag in each file, moving on to the next if there's an error
 	for fileName := range files {
-		logrus.Infof("Attempting bootstrap from %s ...", fileName)
-		img, err := tarball.ImageFromPath(fileName, &archTag)
+		img, err := tarball.ImageFromPath(fileName, &imageTag)
 		if err != nil {
+			logrus.Debugf("Did not find %q in %q: %s", imageName, fileName, err)
 			continue
 		}
+		logrus.Debugf("Found %q in %q", imageName, fileName)
 		return img, nil
-
 	}
+	logrus.Debugf("No local image available for %q: not found in any file in %q", imageName, imagesDir)
 	return nil, nil
 }
+
+// setSystemDefaultRegistry scans the directory at manifestDir. It attempts to load all manifests
+// in that directory as HelmCharts. Any manifests that contain a HelmChart are modified to
+// pass through the systemDefaultRegistry setting to both the Helm job and the chart values.
+// NOTE: This will probably fail if any manifest contains multiple documents. This should
+// not matter for any of our packaged components, but may prevent this from working on user manifests.
+func setSystemDefaultRegistry(manifestsDir string, systemDefaultRegistry string) error {
+	files := map[string]os.FileInfo{}
+	if err := filepath.Walk(manifestsDir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		switch {
+		case info.IsDir():
+			return nil
+		case strings.HasSuffix(path, ".json"):
```

Comment:
I copied this from the deploy controller, but the end result here is that we end up rewriting json files to yaml due to the serializer being hardcoded to output yaml. If this is a problem I can just drop support for json since we're only shipping yaml anyway.